File: backend/src/indexer/solana/decode.ts
import bs58 from 'bs58';
import type { RosterPet } from '@repositories/roster.repository';

/**
 * Decode a raw on-chain `PetAccount` into a roster row. Ported from the old
 * Substreams `map_pets` decoder (backend/graph/solana/substreams/src/lib.rs) so
 * the byte layout stays identical — only the source changed (Helius RPC instead
 * of a Firehose stream).
 *
 * If `PetAccount` in
 * `contracts/solana/cryptopets/programs/cryptopets/src/state.rs` changes, update
 * the discriminator, `PET_ACCOUNT_LEN`, and the offsets below.
 */

/** Anchor discriminator for `PetAccount` — sha256("account:PetAccount")[..8]. */
const PET_ACCOUNT_DISCRIMINATOR = Buffer.from([223, 222, 129, 89, 70, 231, 141, 184]);

/**
 * Byte length of a serialized `PetAccount` — Anchor `PetAccount::SPACE`
 * (8-byte discriminator + Borsh-packed fields). Used as a cheap `dataSize`
 * pre-filter on `getProgramAccounts`.
 */
export const PET_ACCOUNT_LEN = 101;

/** Base58 of the discriminator — for the `getProgramAccounts` memcmp filter. */
export const PET_ACCOUNT_DISCRIMINATOR_B58 = bs58.encode(PET_ACCOUNT_DISCRIMINATOR);

/**
 * Decode account data (including the 8-byte Anchor discriminator). Returns
 * `null` when the bytes are not a `PetAccount`, so callers can safely pass any
 * account touched by a transaction and keep only the ones that decode.
 *
 * Borsh layout after the discriminator, in `state.rs` declaration order:
 *   id u32 | owner [u8;32] | dna u64 | rarity u8 | level u16 | ready_time i64 |
 *   win_count u16 | loss_count u16 | bump u8 | name [u8;32] | name_len u8.
 */
export function decodePetAccount(data: Buffer): RosterPet | null {
    if (data.length !== PET_ACCOUNT_LEN) return null;
    if (!data.subarray(0, 8).equals(PET_ACCOUNT_DISCRIMINATOR)) return null;

    const body = data.subarray(8);

    const id = body.readUInt32LE(0);
    const owner = bs58.encode(body.subarray(4, 36)); // base58 pubkey (matches auth storageKey)
    const dna = body.readBigUInt64LE(36);
    const rarity = body.readUInt8(44);
    const level = body.readUInt16LE(45);
    const readyTime = body.readBigInt64LE(47);
    const winCount = body.readUInt16LE(55);
    const lossCount = body.readUInt16LE(57);
    // body[59] = bump (not indexed)
    const nameBytes = body.subarray(60, 92);
    const nameLen = Math.min(body.readUInt8(92), nameBytes.length);
    const name = nameBytes.subarray(0, nameLen).toString('utf8');

    return {
        chain: 'solana',
        petId: id.toString(),
        owner,
        name,
        dna: dna.toString(),
        level,
        rarity,
        winCount,
        lossCount,
        readyAt: readyTime,
    };
}

File: services/indexer-go/internal/solana/decode.go
package solana

// Decoder for the on-chain shape the adapter consumes:
//   - PetAccount state (port of backend/indexing/solana/scanner/decode.ts)

import (
	"bytes"
	_ "embed"
	"fmt"

	"github.com/radcrew/do-not-stop/services/indexer-go/internal/indexer"
)

//go:embed idl/cryptopets.json
var idlJSON []byte

func resolvePetLayout() (*accountLayout, error) {
	layout, err := resolveAccountLayout(idlJSON, "PetAccount")
	if err != nil {
		return nil, fmt.Errorf("solana: embedded IDL invalid: %w", err)
	}
	return layout, nil
}

// decodePetAccount decodes raw account data (including the 8-byte
// discriminator) into a RosterUpdate with Version unset — the caller stamps
// the slot. Returns ok=false when the bytes are not a PetAccount, so callers
// can pass any account a transaction touched and keep only the ones that
// decode.
func decodePetAccount(layout *accountLayout, data []byte) (indexer.RosterUpdate, bool) {
	if len(data) != layout.totalLen() {
		return indexer.RosterUpdate{}, false
	}
	if !bytes.Equal(data[:8], layout.discriminator) {
		return indexer.RosterUpdate{}, false
	}

	fields, err := decodeStruct(layout.fields, data[8:])
	if err != nil {
		return indexer.RosterUpdate{}, false
	}

	r := newFieldReader(fields)

	nameBuf := r.bytes("name")
	nameLen := r.u64("nameLen")
	// A name length past the fixed buffer means the layout drifted — bail
	// rather than emit a garbage name. Checked before the slice, and only once
	// the two reads above are known to have found their fields.
	if !r.ok() || int(nameLen) > len(nameBuf) {
		return indexer.RosterUpdate{}, false
	}

	update := indexer.RosterUpdate{
		Chain:     "solana",
		PetID:     r.decimal("id"),
		Owner:     r.str("owner"), // base58, no normalization for Solana
		Name:      string(nameBuf[:nameLen]),
		Level:     r.u32("level"),
		Rarity:    r.u32("rarity"),
		DNA:       r.decimal("dna"),
		WinCount:  r.u32("winCount"),
		LossCount: r.u32("lossCount"),
		ReadyAt:   r.i64("readyTime"),

		// v2 fields (state.rs PetAccount). Decoded by IDL name; "0"/0 zero
		// values mean none, matching the on-chain semantics.
		XP:           r.u32("xp"),
		Generation:   r.u32("generation"),
		Parent1ID:    r.decimal("parent1Id"),
		Parent2ID:    r.decimal("parent2Id"),
		BreedCount:   r.u32("breedCount"),
		SpeciesID:    r.u32("speciesId"),
		SpouseID:     r.decimal("spouseId"),
		BreedReadyAt: r.i64("breedReadyTime"),
		TrainReadyAt: r.i64("trainReadyTime"),
		Asset:        r.str("asset"), // base58 Core asset pubkey
	}

	// One check for the whole struct. A field decoded as the wrong type means the IDL and
	// the on-chain account disagree, which is what an account-version bump does; that has
	// to read as "not this account" rather than take the process down mid-subscription.
	if !r.ok() {
		return indexer.RosterUpdate{}, false
	}
	return update, true
}
